WordPress/wordpress-playground
# Cache Playground assets to enable offline support

## Motivation for the change, related issues

The promise of Playground is WordPress anywhere. 
Playground can run in the browser without a server, but it still requires the internet on every load.

This PR adds caching of Playground assets and ensures they are available when offline.

## Implementation details

The Remote Service Worker will start caching when initialized and clean previous versions if they exist.

The WorkerCache class uses browser cache to preload data when the service worker is loaded.
On each unscoped fetch request, it will attempt to load data from the cache by using the `cachedFetch` method.

Only URLs that aren't PHP scripts (end with _.php_) and valid domains will be cached. (see `isValidHostname` for details)

## Testing Instructions (or ideally a Blueprint)

- Follow the [Test offline support instructions](https://github.com/WordPress/wordpress-playground/pull/1535/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R134)
- If you previously loaded the site unregister the service worker and clear cache storage
- After loading Playground **only once** in your browser turn off the server
- In Dev tools > Application > Service Worker set it to Offline (this ensures remote assets are loaded)
- Refresh the page
- Confirm that Playground loaded
- Browse the site, go to WP-admin, make changes...
- It should all work as usual

**Note: In Chrome the offline checkbox sometimes stops working. When this happens Playground will load without remote assets and admin styles (including the admin bar) will look broken. If it happens you can confirm it by running `window.navigator.onLine`. To resolve it uncheck and check the Offline checkbox and reload.**
